### ISSUE TITLE
joint_tests test for interactions not in model (in the backend)

### ISSUE THREAD

**Issue by StaffanBetner** [rvlenth/emmeans#265]
When debugging and trying to find out why joint_tests gets slower with larger models (as in more regressors), I saw that joint_tests calculates contrasts on interactions not in the model. This slows down the running time heavily.

**Comment by rvlenth** [comment]
This is not true. Most **emmeans** functions, including `joint_tests()`, construct a "reference grid" consisting of all factor combinations (see [`vignette("basics", "emmeans")`](https://cran.r-project.org/web/packages/emmeans/vignettes/basics.html)) -- maybe that is what you are saying. However, they do not mess with the model itself, that is left alone.

I am guessing that the slow performance you are experiencing is happening in the construction of the reference grid, not in `joint_tests()` itself. Because of that, I suggest constructing the reference grid separately. This will not save any processing time, but it keeps you from having to do it more than once. So, instead of `joint_tests(model)`, I suggest
```r
RG <- ref_grid(model)
joint_tests(RG)
```
This will also clarify where the computational lags are occurring.

You don't say what kind of models you are using, but the ones most susceptible to slow calculations seem to be ones fitted using `lme4::lmer()`, `nlme::lme()`, and `nlme::gls()`. In those cases, the performance issues come up in degrees-of-freedom calculations. For `lmer` models, the default is to use Kenward-Roger d.f. (using routines in the **pbkrtest** package), and that involves calculating an adjusted covariance matrix, which is very intensive for large numbers of predictors. In `gls` and `lme`, an approximate Satterthwaite method is used whereby the model is re-fitted several times with slight perturbations in the response variable -- again, that can take a noticeable amount of time. 

You can boost the performance considerably for those models by selecting simpler d.f. methods. See [`vignette("models", "emmeans")`](https://cran.r-project.org/web/packages/emmeans/vignettes/models.html) for details. This comes at a price of using less refined methods, but if the d.f. are going to be large anyway, that will not make much difference. Here are some alternatives:
```r
### For lmer models
ref_grid(model, mode = "satterth")   ## somewhat faster
ref_grid(model,  mode = "asymp")    ## quite a bit faster

### For lme models
ref_grid(model, mode = "contain")

### For gls models
ref_grid(model, mode = "df.error")
```
Again, I think you are going to find that the computing issues come up with constructing the reference grid, not running `joint_tests()`. It is true that the latter does involve some facy matrix calculations, but they are pretty routine linear-algebra procedures.

**Comment by rvlenth** [comment]
Oops, you are right. It's been a long time since I wrote this code, but apparently the most expeditious thing to do (programming-wise) was to just look at all factor combinations. I still don't think this slows down the computation all that much. Do you have a reproducible example where you see a noticeable issue with this?

**Comment by rvlenth** [comment]
I did put in a little code that looks at `attr(terms, "factors")` to determine which terms are actually in the model. It's a little tricky because I have to convert things like `factor(dose)` to just `dose`. But this does avoid testing contrasts that are not in the model, which may save some time for models with lots of factors and not many interaction terms.

**Comment by rvlenth** [comment]
The "fix" in `cb576e2` created a new bug, because I didn't consider `emmGrid` objects not directly created by `ref_grid()`. They may not have the `terms` attribute at all, or there may be additional factors not included in `terms`. One solution (tempting) is to just restore it to the way it was before. But for now at least, I decided to create extra rows in my reference table to account for them. Just in case you're interested, here is a test run
```r
> pigs.add = lm(conc ~ source + factor(percent), data = pigs)
> RGA = ref_grid(pigs.add)
> joint_tests(RGA, show0df = TRUE)
 model term df1 df2 F.ratio p.value
 source       2  23  21.660 <.0001 
 percent      3  23   6.038 0.0035 
```
With `show0df = TRUE`, **emmeans** 1.5.4 would have shown a 0-df row for `source:percent`.

```r
> # Create a new factor 'type' that nests 'source'
> RGAN = add_grouping(RGA, "type", "source", c("animal", "vegetable", "animal"))

> joint_tests(RGAN, show0df = TRUE)
 model term   df1 df2 F.ratio p.value note
 source         2  23  21.660 <.0001   d  
 percent        3  23   6.038 0.0035      
 type           1  23   0.701 0.4109      
 source:type    2  23  21.660 <.0001   d  
 percent:type   0  NA  nonEst     NA   d e

d: df1 reduced due to linear dependence 
e: df1 reduced due to non-estimability 
```
The `d` flags result from the fact that `source` is nested in `type`.  This time, we do test two unnecessary cases. One is `source:type` which is the same as `source`, and the other is `percent:type` which is unnecessary because it would have been 1 d.f. of `source:percent` which is not in the model. It is conceivable that `add_grouping()` could be made mode sophisticated to take care of such situations, but that's for another day.

**Comment by rvlenth** [comment]
I am closing this issue as I think it has been resolved, and the revised code is now incorporated in emmeans 1.5.5-1
